### DISCUSSION TITLE
When will the Rust API be ready for production use?

### DISCUSSION THREAD

**Discussion by JustinWickMS** [microsoft/onnxruntime#15083] [Other Q&A]
I'm starting a greenfield project here at Microsoft using ONNX, and we intend to deploy to production later this year. I would like to use Rust and the Rust API for ONNX, is there any notion of when the Rust API will be stable, performant, and reliable enough to be used for real customers?

Instead of using the Rust interface, is the C or C++ interface best to call from Rust in the meantime?

**Comment by pranavsharma** [comment]
Rust API is here https://github.com/microsoft/onnxruntime/tree/main/rust. The more users we've the more it gets tested, stable and reliable.

**Reply by JustinWickMS** [reply]
Yes, I've seen that, however I was unable to determine which ONNX functionality was implemented, which of that is tested and known to work, how stable the rust interface is going to be in the next year, and if there are any critical limitations.

Does anyone have answers to that?

**Comment by asyncth** [comment]
You can also use [ort](https://github.com/pykeio/ort) Rust crate in the meantime.

**Comment by JustinWickMS** [comment]
Can anyone answer my question above? Is there someone I can contact to figure out the answer?

**Reply by wschin** [reply]
@RyanUnderhill, any comments? I saw you reviewed rust binding PR. I guess @JustinWickMS is asking for something like a supported table.

**Reply by RyanUnderhill** [reply]
> @RyanUnderhill, any comments? I saw you reviewed rust binding PR. I guess @JustinWickMS is asking for something like a supported table.

I pinged the primary Rust contributor, to see if they have any suggestions. I'd say currently it's still experimental, but the only way to stabilize is to use it and fix any rough edges.

**Reply by boydjohnson** [reply]
> Can anyone answer my question above? Is there someone I can contact to figure out the answer?

@JustinWickMS I would consider the Rust bindings as being experimental. I know of an effort to alter them to be more performant (https://github.com/nbigaouette/onnxruntime-rs/issues/112#issuecomment-1504084389). So I would expect large changes coming in a PR soon. 

It is hard to answer the question of when for "stable, performant, and reliable".

**Reply by boydjohnson** [reply]
@wschin @RyanUnderhill I think a supported table is something that we could provide here. @RyanUnderhill and I could confer on the common set of features and produce a table.

**Reply by jackos** [reply]
@JustinWickMS I'm using the [ort](https://github.com/pykeio/ort) crate in production and it's working great, don't use the crates in this repository they're very outdated and ort was forked from them and updated for 14.1.0 with a bunch of extra improvements. Twitter is also using `ort` for their recommendations engine according to the Readme.
